Retry just-in-time translation when setting custom path

swissspidy opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

This is a follow-up to https://core.trac.wordpress.org/changeset/59157.

```
I18N: Retry just-in-time translation when setting custom path.

This is a follow-up to [59127] and [59157], which changed the `load_*_textdomain()` functions to reuse the existing just-in-time translation loading functionality. While this works great for cases where those functions are called too early, it broke cases where they are called after the just-in-time logic already ran.

To solve this, cached entries from the just-in-time logic are reset, so that the loading will be simply reattempted the next time a translation is attempted (e.g. by calling `__()`).

Props swissspidy, peterwilsoncc, twvania.
Fixes #62337.
See #44937
```

Trac ticket: https://core.trac.wordpress.org/ticket/62337

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

swissspidy committed:
Retry just-in-time translation when setting custom path

swissspidy committed:
Add unit test

peterwilsoncc commented on tests/phpunit/tests/l10n/loadTextdomainJustInTime.php:
```diff
@@ -342,4 +342,36 @@ public function test_get_locale_is_called_only_once_per_textdomain_with_custom_l
+		$is_textdomain_loaded_after = is_textdomain_loaded( 'custom-internationalized-plugin' );
+
+		remove_filter( 'locale', array( $this, 'filter_set_locale_to_german' ) );
+
```
Reset during teardown.

```suggestion
```

peterwilsoncc reviewed (commented):
This tests well using the sample theme provided by @t-hamano.

I did a revert of the source changes (since force pushed out of existence) to validate the tests and they [fail on trunk as expected](https://github.com/WordPress/wordpress-develop/actions/runs/11693508463?pr=7713). 

It would be good to get another set of eyes from either @SergeyBiryukov or @ocean90 as they have a deeper knowledge of the feature.

One nit pick inline

swissspidy committed:
Update tests/phpunit/tests/l10n/loadTextdomainJustInTime.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

swissspidy committed:
Remove en_US exclusion

Custom themes/plugins with only en_US files should have their translations loaded as well

swissspidy committed:
Merge branch 'trunk' into fix/62337-l10n-jit

swissspidy committed:
Fix test

swissspidy committed:
Merge branch 'trunk' into fix/62337-l10n-jit

desrosj commented on src/wp-includes/class-wp-textdomain-registry.php:
```diff
@@ -153,6 +153,16 @@ public function set( $domain, $locale, $path ) {
 	 */
 	public function set_custom_path( $domain, $path ) {
+		// If just-in-time loading was triggered before, reset the entry so it can be tried again.
+
```
```suggestion
```

desrosj reviewed (approved):
I've given this some good testing today and I'm seeing all of the scenarios described in Trac as resolved with the exception of the most recently raised edge case: `My plugin lets users override GlotPress translations with their own custom translation files`.

Would also appreciate another review from someone with deeper l10n/i18n knowledge, but I think this one is probably ready.

apermo reviewed (approved):
Lgtm

SergeyBiryukov reviewed (approved):
Thanks for the PR, looks good to me 👍

desrosj commented:
Merged in https://core.trac.wordpress.org/changeset/59430.
